### PR TITLE
[Commands] Add the possibility to register commands to specific guilds

### PR DESCRIPTION
**Pull request by Lounek09**
Implement #1734 by adding overwrite method for all the `CommandExtension.AddCommand` and `CommandExtension.AddCommands` like DSP.SlashCommands does with `SlashCommandsExtension.RegisterCommands`

### CODE REVIEW

File: DSharpPlus.Commands/DefaultCommandExecutor.cs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+namespace DSharpPlus.Commands;
```
Comment (Plerx2493) [inline DSharpPlus.Commands/DefaultCommandExecutor.cs:1]: Namespace blow usings please

File: DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs [hunk 2]
```diff
@@ -19,7 +18,6 @@ namespace DSharpPlus.Commands.Processors.SlashCommands;
 using DSharpPlus.Commands.Trees.Metadata;
 using DSharpPlus.Entities;
 using DSharpPlus.EventArgs;
```
Comment (Plerx2493) [inline DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs:20]: nit: old stlye with linebreaks prefered, atleast personally

**Review by Plerx2493 — COMMENTED** [review]
Generall LGTM, some minor comments

**Review by OoLunar — CHANGES_REQUESTED** [review]
Could we also do a `RequireGuildIdCheckAttribute`?

**Comment by akiraveliara** [comment]
wrt RequireGuildIdCheck: users can just register commands to specifically that guild, no

**Comment by Plerx2493** [comment]
Question, how are we handling 2 Commands with the same name but other guilds? e.g. a `/rule` command could have different implementations for different guilds

**Comment by OoLunar** [comment]
> wrt RequireGuildIdCheck: users can just register commands to specifically that guild, no

They could but I personally register all my commands by passing the assembly. I'd rather use attributes instead of filtering out the types that need to be registered to specific guilds.

**Comment by akiraveliara** [comment]
i have opinions on encouraging such a pattern, but at the very least it's not a check then but registration metadata

File: DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs [hunk 3]
```diff
@@ -191,12 +190,40 @@ public async Task RegisterSlashCommandsAsync(CommandsExtension extension)
...
+            }
+
+            globalApplicationCommands.Add(await this.ToApplicationCommandAsync(command));
+        }
+
+        List<DiscordApplicationCommand> discordCommands = [];
+        if (extension.DebugGuildId is 0)
+        {
+            discordCommands.AddRange(await extension.Client.BulkOverwriteGuildApplicationCommandsAsync(extension.DebugGuildId, globalApplicationCommands));
+            discordCommands.AddRange(await extension.Client.BulkOverwriteGuildApplicationCommandsAsync(extension.DebugGuildId, guildsApplicationCommands.SelectMany(x => x.Value)));
```
[the remark this answers is not in the document]
Comment (Lounek09) [reply DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs]: I don't understand, this doesn't compile 😭

[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs]: Yeah I was in the wrong here. Carry on lol

**Comment by Lounek09** [comment]
> Could we also do a `RequireGuildIdCheckAttribute`?

done !

**Review by akiraveliara — COMMENTED** [review]
consider embedding the guild ids in upload metadata rather than a check, that way they aren't at all visible if the check would fail

**Comment by Lounek09** [comment]
> consider embedding the guild ids in upload metadata rather than a check, that way they aren't at all visible if the check would fail

Yeah, that's better you're right (and it's done🫣)

**Review by akiraveliara — APPROVED** [review]

File: DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs [hunk 4]
```diff
@@ -180,23 +178,55 @@ public async ValueTask ClearDiscordSlashCommandsAsync(bool clearGuildCommands =
...
+                discordCommands.AddRange(await extension.Client.BulkOverwriteGuildApplicationCommandsAsync(guildId, guildCommands));
+            }
+        }
+        else
+        {
+            discordCommands.AddRange(await extension.Client.BulkOverwriteGuildApplicationCommandsAsync(extension.DebugGuildId, globalApplicationCommands));
+            discordCommands.AddRange(await extension.Client.BulkOverwriteGuildApplicationCommandsAsync(extension.DebugGuildId, guildsApplicationCommands
+                .SelectMany(x => x.Value)
+                .GroupBy(x => x.Name)
+                .Select(x => x.First())));
```
Comment (OoLunar) [inline DSharpPlus.Commands/Processors/SlashCommands/SlashCommandProcessor.cs:228]: What's this LINQ query for? Could we add a code comment for this?

File: DSharpPlus.Commands/RegisterToGuildsAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method | AttributeTargets.Delegate)]
+public class RegisterToGuildsAttribute : Attribute
+{
+    /// <summary>
+    /// The guild ids to register this command to.
+    /// </summary>
+    public IReadOnlyList<ulong> GuildIds { get; init; }
+
+
```
Comment (OoLunar) [inline DSharpPlus.Commands/RegisterToGuildsAttribute.cs:14]: Double newline

File: DSharpPlus.Commands/RegisterToGuildsAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    public RegisterToGuildsAttribute(params ulong[] guildIds)
+    {
+        if (guildIds.Length == 0)
+        {
+            throw new ArgumentException("You must provide at least one guild ID.");
+        }
+
+        this.GuildIds = guildIds;
+    }
+
```
Comment (OoLunar) [inline DSharpPlus.Commands/RegisterToGuildsAttribute.cs:28]: Trailing newline

**Review by OoLunar — CHANGES_REQUESTED** [review]

**Review by OoLunar — APPROVED** [review]
LGTM. Tested?

**Comment by OoLunar** [comment]
Oh, does it work for text commands too?

**Comment by akiraveliara** [comment]
they mentioned running this a few times, so

**Comment by Lounek09** [comment]
> LGTM. Tested?

Yup, I run it in production right now, that's how I found the last bug 😅

> Oh, does it work for text commands too?

See this change : https://github.com/DSharpPlus/DSharpPlus/pull/1854/files#diff-89f69a773b7a1cf1cd35141ebb4b1affe6273067dfb3ad40a0119a54b8b9e75c

**Pull request merged**